stock_dynamic_routing/models/stock_routing.py

```diff
@@ -0,0 +1,262 @@
+        candidate_rules = self.env["stock.routing.rule"].search(
+            [
+                "|",
+                "&",
+                "&",
+                ("routing_location_id", "in", pull_location_tree.ids),
+                ("routing_picking_type_id", "=", picking_type.id),
+                ("method", "=", "pull"),
+                "&",
+                "&",
+                ("routing_location_id", "in", push_location_tree.ids),
+                ("routing_picking_type_id", "=", picking_type.id),
+                ("method", "=", "push"),
+            ]
+        )
+        candidate_rules.sorted(lambda r: (r.routing_id.sequence, r.sequence))
+        rule = self._get_location_routing_rule(
+            move, pull_location_tree, candidate_rules
+        )
+        if rule:
+            return rule
+
+        rule = self._get_location_routing_rule(
+            move, push_location_tree, candidate_rules
+        )
+        if rule:
+            return rule
+
+        empty_rule = self.env["stock.routing.rule"].browse()
+        return empty_rule
+
+    def _routing_rule_for_move_lines(self, moves):
+        """Return a routing rule for move lines
+
+        Look first for a pull routing rule, if no match, look for a push
+        routing rule.
+
+        :param move: recordset of the move
+        :return: dict {move: {rule: move_lines}}
+        """
+        # ensure the cache is clean
+        self.__cached_is_rule_valid_for_move.cache_clear()
+
+        result = {
+            move: defaultdict(self.env["stock.move.line"].browse) for move in moves
+        }
+        for move_line in moves.mapped("move_line_ids"):
+            rule = self._find_rule_for_location(
+                move_line.move_id, move_line.location_id, move_line.location_dest_id
+            )
+            result[move_line.move_id][rule] |= move_line
+
+        # free memory used for the cache
+        self.__cached_is_rule_valid_for_move.cache_clear()
+        return result
+
+    def _routing_rule_for_moves(self, moves):
+        """Return a routing rule for moves
+
+        Look first for a pull routing rule, if no match, look for a push
+        routing rule.
+
+        :param move: recordset of the move
+        :return: dict {move: rule}}
+        """
+        # ensure the cache is clean
+        self.__cached_is_rule_valid_for_move.cache_clear()
+
+        result = {}
+        for move in moves:
+            rule = self._find_rule_for_location(
+                move, move.location_id, move.location_dest_id
+            )
+            result[move] = rule
+
+        # free memory used for the cache
+        self.__cached_is_rule_valid_for_move.cache_clear()
+        return result
+
+    # Do not use ormcache, which would invalidate cache of other workers every
+    # time we clear it. We only need a local cache used for the duration of the
+    # execution of
+    @lru_cache()
+    def __cached_is_rule_valid_for_move(self, rule, move):
```

Comment:
why double underscore?

Reply:
I want to use a method for the `@lru_cache`, but in practice this method is really an internal thing of the algorithm, not meant to be overloaded in any way. Using double underscore makes it hard to overload (you would have to use `_StockRouting__cached_is_rule_valid_for_move` IIRC). I should write it in the docstring.

Reply:
yeah, that was the question: why making hard to override? :+1: